### PR TITLE
TV Android app - Add commissioning User Prompter to TV android app

### PR DESCRIPTION
**Pull request by lazarkov**
#### Problem
- TV Android App does not have a UI support for [UserPrompter logic](https://github.com/project-chip/connectedhomeip/blob/master/examples/tv-app/android/java/AppImpl.cpp#L50-L100)
- End users are not able to confirm or decline commissioning request
- End users are not able to enter PIN to confirm commissioning request

#### Change overview
- Added JNI layer for Native C++ UserPrompter logic so that code can prompt end users with UI dialog 
- Added JNI layer to Android UserPrompter logic so that code can pass back user response (accept/decline) or pin code

#### Testing
Manually tested the flow using the instrucitons @chrisdecenzo shared with me:
1) Overwrite your src/platform/android/java/chip/platform/NsdManagerServiceResolver.java with the file provided
2) For full shell command functionality from casting app, replace your OperationalDeviceProxy.h/cpp with the file
3) Create a UDP port mapping from 5750 to 5650
telnet localhost 5554
auth XXXXX
redir add udp:5750:5650
redir add udp:5740:5640
4) run linux tv-casting-app. It won't discover the TV so you need to manually force a user-directed-commissioning request to the TV using the remapped port:
cast sendudc 127.0.0.1 5750
5) The TV should receive the UDC command and automatically commission the casting app. Watch tv-app logs for messages like:
------PROMPT USER: is requesting permission to cast to this TV, approve? [0x0000_0000,0x0000_0000,AB4E0808D7EFD3B3]
------PROMPT USER: please enter pin displayed in casting app
------PROMPT USER: commissioning success
Right now, tv-app is hardcoded to automatically commission since there is no current dialog box implemented.

### CODE REVIEW

**Comment by lazarkov** [comment]
Here's the NsdManagerServiceResolver code.
[NsdManagerServiceResolver.txt](https://github.com/project-chip/connectedhomeip/files/8794424/NsdManagerServiceResolver.txt)

**Comment by lazarkov** [comment]
Uploading FirstUiDemoOfCommissioning.mov…

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        .create()
+        .show();
+  }
+
+  public void promptCommissioningSucceeded(int vendorId, int productId, String commissioneeName) {
+    NotificationCompat.Builder builder =
+        new NotificationCompat.Builder(activity, CHANNEL_ID)
+            .setSmallIcon(R.drawable.ic_baseline_check_24)
+            .setContentTitle("Device connected")
+            .setContentText("Your device " + commissioneeName + " is ready for use.")
```
Comment (chrisdecenzo) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:81]:
```suggestion
            .setContentText("Success. " + commissioneeName + " can now cast to this device. Visit settings to manage access control for casting.")
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            })
+        .create()
+        .show();
+  }
+
+  public void promptCommissioningSucceeded(int vendorId, int productId, String commissioneeName) {
+    NotificationCompat.Builder builder =
+        new NotificationCompat.Builder(activity, CHANNEL_ID)
+            .setSmallIcon(R.drawable.ic_baseline_check_24)
+            .setContentTitle("Device connected")
```
Comment (chrisdecenzo) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:80]:
```suggestion
            .setContentTitle("Connection Complete")
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            .setPriority(NotificationCompat.PRIORITY_DEFAULT);
+
+    notificationManager.notify(SUCCESS_ID, builder.build());
+  }
+
+  public void promptCommissioningFailed(String commissioneeName, String error) {
+    NotificationCompat.Builder builder =
+        new NotificationCompat.Builder(activity, CHANNEL_ID)
+            .setSmallIcon(R.drawable.ic_baseline_clear_24)
+            .setContentTitle("Error occurred")
```
Comment (chrisdecenzo) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:91]:
```suggestion
            .setContentTitle("Connection Failed")
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            .setContentText("Device " + commissioneeName + " experienced error: " + error)
```
Comment (chrisdecenzo) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:92]:
```suggestion
            .setContentText("Failed. " + commissioneeName + " experienced error: " + error + ".")
```

**Review by chrisdecenzo — APPROVED** [review]
Please see UX text changes

File: examples/tv-app/android/java/src/com/tcl/chip/tvapp/UserPrompterResolver.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/*
+ *   Copyright (c) 2021 Project CHIP Authors
```
Comment (lazarkov) [inline examples/tv-app/android/java/src/com/tcl/chip/tvapp/UserPrompterResolver.java:2]:
```suggestion
 *   Copyright (c) 2022 Project CHIP Authors
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MainActivity.java [hunk 6]
```diff
@@ -40,6 +41,8 @@ protected void onCreate(Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
     setContentView(R.layout.activity_main);
 
+    MatterServant.get().setActivity(this);
```
Comment (lazarkov) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MainActivity.java:44]:
```suggestion
    // MainActivity is needed to launch dialog prompt
    // in UserPrompter
    MatterServant.get().setActivity(this);
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
+package com.matter.tv.server;
+
+import static androidx.core.content.ContextCompat.getSystemService;
+
+import android.app.Activity;
+import android.app.NotificationChannel;
+import android.app.NotificationManager;
+import android.os.Build;
+import android.widget.EditText;
```
Comment (lazarkov) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:9]:
```suggestion
import android.util.Log;
import android.widget.EditText;
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+  public MatterCommissioningPrompter(Activity activity) {
+    this.activity = activity;
+    this.createNotificationChannel();
+  }
+
+  public void promptForCommissionOkPermission(
+      int vendorId, int productId, String commissioneeName) {
+    // TODO: find app by vendorId and productId
+    AlertDialog.Builder builder = new AlertDialog.Builder(activity);
```
Comment (lazarkov) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:31]:
```suggestion
    Log.d(TAG, "Received prompt for OK permission vendor id:" + vendorId + " productId:" + productId + ". Commissionee: " + commissioneeName);
    AlertDialog.Builder builder = new AlertDialog.Builder(activity);
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+              OnPromptDeclined();
+            })
+        .create()
+        .show();
+  }
+
+  @Override
+  public void promptForCommissionPinCode(int vendorId, int productId, String commissioneeName) {
+    // TODO: find app by vendorId and productId
+
```
Comment (lazarkov) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:53]:
```suggestion
    Log.d(TAG, "Received prompt for PIN code vendor id:" + vendorId + " productId:" + productId + ". Commissionee: " + commissioneeName);
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            "Cancel",
+            (dialog, which) -> {
+              OnPinCodeDeclined();
+            })
+        .create()
+        .show();
+  }
+
+  public void promptCommissioningSucceeded(int vendorId, int productId, String commissioneeName) {
+    NotificationCompat.Builder builder =
```
Comment (lazarkov) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:77]:
```suggestion
    Log.d(TAG, "Received prompt for success vendor id:" + vendorId + " productId:" + productId + ". Commissionee: " + commissioneeName);
    NotificationCompat.Builder builder =
```

File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                "Success. "
+                    + commissioneeName
+                    + " can now cast to this device. Visit settings to manage access control for casting.")
+            .setPriority(NotificationCompat.PRIORITY_DEFAULT);
+
+    notificationManager.notify(SUCCESS_ID, builder.build());
+  }
+
+  public void promptCommissioningFailed(String commissioneeName, String error) {
+    NotificationCompat.Builder builder =
```
Comment (lazarkov) [inline examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MatterCommissioningPrompter.java:91]:
```suggestion
Log.d(TAG, "Received prompt for failure vendor id:" + vendorId + " productId:" + productId + ". Commissionee: " + commissioneeName);
    NotificationCompat.Builder builder =
```

**Review by hawk248 — APPROVED** [review]

**Pull request merged**